src/pymedphys/dicom/_level2/anonymise.py

```diff
@@ -213,6 +239,36 @@ def anonymise_dicom_dataset(ds, replace_values=True, delete_private_tags=True,
     return ds_anon
 
 
+def anonymise_dicom_file(dicom_filepath, overwrite_file=False, replace_values=True,
+                         delete_private_tags=True, keywords_to_keep=None,
+                         ignore_unknown_tags=False, delete_unknown_tags=False):
+
+    ds = pydicom.dcmread(dicom_filepath)
+
+    if overwrite_file:
+        dicom_anon_filepath = dicom_filepath
+    else:
+        dicom_anon_filepath = pjoin(dirname(dicom_filepath),
+                                    'Anonymised_' + basename(dicom_filepath))
```

Comment:
Keen for thoughts on this. Currently, if the user doesn't want to replace the file, this function just prepends 'Anonymized' to the existing filename.

One problem with this approach is that the filename itself can have identifying information. Although many DICOM filenames tend to just consist of the SOPInstanceUID preprended with something like 'CT' or 'RP' it's also not so uncommon for DICOM filenames to consist of patient info. E.g. ARIA exports DICOM files containing PatientID by default.

Should we try to handle this, or not? E.g., we could search the filename for instances of, say, PatientID and Patient Name - heck, maybe even search for all of the anonymized DICOM tag values in the filename - and either wipe them or replace them with dummy values. The trade-off as I see it is:

- If we do this, the user can more completely rely on the function's anonymisation, BUT

- It introduces a level of inconvenience to the user, where it'll be harder to see which of the original files a new anonymized file corresponds to. This would especially be annoying for anonymisation within a folder.

Is it enough to simply warn the user that the filename is untouched and that the user should take care to rename themselves? Or maybe even provide both solutions?

Reply:
I think we should create a family of anonymise CLI functions. One should be of the form:

```
pymedphys dicom anonymise-file input_file output_file
```

Which completely leaves the filename up the user.

Another:

```
pymedphys dicom anonymise-directory input_dir output_dir
```

Which recursively walks through a directory, anonymises the DICOM files and copies them into the new directory with the same names and the same directory tree structure.

I think from there we need to treat the user like a responsible adult. We are anonymising the DICOM header, nothing more. Not the image data, not the file names. I don't even believe a user warning is needed.

What do you think? I think we need to adopt a "assume user is responsible adult" stance, otherwise our interfaces may become very convoluted and the logic within our functions over complicated beyond doing the task at hand.

We could have a third function, not DICOM specific, that searches a range of supported file formats as well as directory and file names, for a string that looks like a name. That function could then be recommended to be run after the DICOM anonymiser within the documentation?

Reply:
> I think we should create a family of anonymise CLI functions. One should be of the form:
> 
> ```
> pymedphys dicom anonymise-file input_file output_file
> ```
> Which completely leaves the filename up the user.
> 
> Another:
> 
> ```
> pymedphys dicom anonymise-directory input_dir output_dir
> ```
> Which recursively walks through a directory, anonymises the DICOM files and copies them into the new directory with the same names and the same directory tree structure.


Or maybe simply just:

```
pymedphys dicom anonymise input_file_dir_or_zip output_file_dir_or_zip
```

I'm not sure...

Reply:
To be honest, I don't like the recursive dir copy with matching structure. It's unnecessarily slow and, IMHO, introduces confusion. What if you have other non-DICOM files in there? Do you copy them across or not? If so, added time. If not, loss of structure that might be important...

Reply:
Fair point. The user story I am trying to address is "I have a collection of DICOM files in a directory tree and I want to anonymise the DICOM files before sending to a third party."

I'm open to not doing it the way I mention but the result of the "anonymisation name append" method is that the user then needs to copy the directory themselves (if they want to keep the non-anonymised version) and then within the copied directory they need to delete all DICOM files that don't have "anonymised", making sure to not miss one.

Reply:
Also when I say "user story" I mean the following:

https://www.mountaingoatsoftware.com/agile/user-stories

Reply:
> It's unnecessarily slow 

How is it any slower?

Reply:
> to copy the directory themselves (if they want to keep the non-anonymised version) and then within the copied directory they need to delete all DICOM files that don't have "anonymised" in the file name, making sure to not miss one.

The user would copy the dir themselves and then anonymise in-place, I think. So no manual deletion necessary. I admit that the manual copy would be nice to avoid though. But then again, I personally would rarely want to maintain the original dir structure... so a manual copy is almost necessary.

I have noticed in my own tinkering that a manual copy is fantastically fast on a Windows machine. Python's `shutil` and `xcopy` are ridiculously slow by comparison when copying recursively. Maybe I haven't been clever enough - please try yourself! I also just found [this SO answer](https://stackoverflow.com/a/28584857/4665458) that might address the issue.

Reply:
Another related but separate point: prepending "anonymous" or similar to the filename clearly flag to the user that the file has actually been anonymised. If we instead maintain the original filename, the user is relying on the dir name to flag anonymisation - which falls apart if they ever copy the file out of that dir...

Of course, they could open the DICOM to inspect it for anonymity. But this is inconvenient - DICOM files are binary, after all.

Reply:
> Python's shutil and xcopy are ridiculously slow by comparison when copying recursively.

I wouldn't imagine using xcopy, instead just output the DICOM files to a the new location. No slower than what is being done to write a new anonymised dicom file with annonymised prepended.

Reply:
> I wouldn't imagine using xcopy, instead just output the DICOM files to a the new location. No slower than what is being done to write a new anonymised dicom file with annonymised prepended.

So not copying any non-DICOM files?

Reply:
Yeah I suppose not... Hmm...

What about this, have it anonymise inplace, don't append anything, overwrite the old files, then provide a "is anonymised" CLI function if someone wants to check if it has been anonymised.

The "is anonymised" could use pydicom's "read header only" flag so that it would run really quick even for large pixel data files. And it then goes and verifies if a file has been anonymised directly, by reading the header.

Some files, like "BFG cleaner" create a log, and they also have a "this is what I am going to do, is that okay" prompt. (A prompt which can be removed with a "no-interaction" flag).

As in, if it is in place, and overwrites old files, then it could print a list of all subfiles it will overwrite before it does anything.

Then once it has done it, it could write out a log of all files it anonymised into the top directory, a date and time of when it did it, and potentially a list of tags that were found that needed modification within the files. If tags were deleted a lost of tags that were deleted.

And then maybe even a link to the PyMedPhys DICOM anonymisation CLI docs at the top of the log so that the third party who receives the files knows that tool was used to undergo the anonymisation...

Reply:
I think an `is_anonymised()` function is worthwhile (e.g. for broad repo checks), but I still really think we ought to simply rename the files if anonymisation is not in place. It really is standard behaviour - every anonymiser I've used does this, including RayStation itself. Pylinac even does it for its log file anonymiser.

- I think your concern at having to delete the originals is solved simply by the user choosing the `in_place` option.

- I think it is far simpler, more expected and more novice-friendly to simply rename the file (with an "anonymous" prepend) for non-in-place anonymisation. It's a very readily visible flag that the file has indeed been anonymised that neither relies on the file remaining in its dir structure, nor requires a CLI run of `is_anonymised()`.